components/tensorboard-controller/controllers/tensorboard_controller.go

```diff
@@ -0,0 +1,230 @@
+// +kubebuilder:rbac:groups=tensorboard.kubeflow.org,resources=tensorboards/status,verbs=get;update;patch
+func (r *TensorboardReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	logger := r.Log.WithValues("tensorboard", req.NamespacedName)
+
+	// your logic here
+	instance := &tensorboardv1alpha1.Tensorboard{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Reconcile k8s deployment.
+	deployment := generateDeployment(instance)
+	if err := ctrl.SetControllerReference(instance, deployment, r.Scheme); err != nil {
+		return ctrl.Result{}, err
+	}
+	if err := reconcilehelper.Deployment(ctx, r, deployment, logger); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Reconcile k8s service.
+	service := generateService(instance)
+	if err := ctrl.SetControllerReference(instance, service, r.Scheme); err != nil {
+		return ctrl.Result{}, err
+	}
+	if err := reconcilehelper.Service(ctx, r, service, logger); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Reconcile istio virtual service.
+	virtualService, err := generateVirtualService(instance)
+	if err := ctrl.SetControllerReference(instance, virtualService, r.Scheme); err != nil {
+		return ctrl.Result{}, err
+	}
+	if err := reconcilehelper.VirtualService(ctx, r, virtualService.GetName(), virtualService.GetNamespace(), virtualService, logger); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *TensorboardReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&tensorboardv1alpha1.Tensorboard{}).
+		Complete(r)
+}
+
+func generateDeployment(tb *tensorboardv1alpha1.Tensorboard) *appsv1.Deployment {
+	return &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      tb.Name,
+			Namespace: tb.Namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: proto.Int32(1),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": tb.Name,
+				},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"app": tb.Name},
+				},
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyAlways,
+					Containers: []v1.Container{
+						{
+							Name:            "tensoroboard",
+							Image:           "tensorflow/tensorflow:1.8.0",
+							ImagePullPolicy: "IfNotPresent",
+							Command:         []string{"/usr/local/bin/tensorboard"},
+							Args: []string{
+								"--logdir=" + tb.Spec.LogsCloudPath, //example: "--logdir=gs://tf_mnist_writebycode1/mnist_tutorial/",
```

Comment:
Same as top comment, since we have a PodTemplateSpec, we should be able to add a PVC in the Spec and reference it here

Reply:
that sounds reasonable, updated to pvc is added only when the logpath is cloudpath(which matches the logic in tensorboard).